File: SDKPackages/da/ios/Samples/KitchenSinkNew/KitchenSinkNew/TagOrderCompletion.m
/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (C) Copyright IBM Corp. 2016
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 ******************************************************************************/

#import "TagOrderCompletion.h"
#import "Product.h"
#import "DigitalAnalytics.h"

@interface TagOrderCompletion()
@property (nonatomic,strong) NSString *pageName;
@property (nonatomic,strong) ShoppingCart *shoppingCart;
@property (nonatomic,strong) Order *order;
@end

@implementation TagOrderCompletion

@synthesize pageName;
@synthesize shoppingCart;
@synthesize order;

-(id) initTag:(NSString *) pageNameParam shoppingCart:(ShoppingCart *)shoppingCartParam order:(Order *)orderParam
{
    self = [super init];
    
    self.pageName = pageNameParam;
    self.shoppingCart = shoppingCartParam;
    self.order = orderParam;
    
    return self;
}

-(void) executeTag
{
    BOOL shopAction9Success = YES;
    
    NSString *orderId = self.order.orderId;
    NSString *orderSubtotal = self.order.subTotal;
    NSString *customerId = self.order.customerId;
    NSString *currencyCode = self.order.currencyCode;
    NSString *shippingCharge = self.order.shippingCharge;
    NSString *customerCity = self.order.customerCity;
    NSString *customerState = self.order.customerState;
    NSString *customerZip = self.order.customerZip;
    NSArray *orderAttributes = self.order.attributes;
    
    //Fire a ShopAction9 for each Product in the SHopping Cart
    NSSet *products = self.shoppingCart.products;
    for(Product* product in products)
    {
        NSString *productId = product.productId;
        NSString *productName = product.name;
        NSString *categoryId = product.categoryId;
        NSString *quantity = [NSString stringWithFormat:@"%d",product.quantity];
        NSString *baseUnitPrice = product.baseUnitPrice;
        NSArray *attributes = product.attributes;
        
        if(![DigitalAnalytics fireShopAction9:productId productName:productName quantity:quantity baseUnitPrice:baseUnitPrice category:categoryId orderId:orderId orderSubTotal:orderSubtotal customerId:customerId currencyCode:currencyCode attributes:attributes])
        {
            shopAction9Success = NO;
        }
    }
    
    //Fire a Tag for Order Placement
    BOOL orderSuccess = [DigitalAnalytics fireOrder:orderId subtotal:orderSubtotal shippingCharge:shippingCharge customerId:customerId customerCity:customerCity customerState:customerState customerZip:customerZip currencyCode:currencyCode attributes:orderAttributes];
    
    //Indicate success
    [self finish:(shopAction9Success && orderSuccess)];
}

@end

File: SDKPackages/da/android/SampleCode/DigitalAnalyticsKitchenSink/app/src/main/java/com/tl/digitalanalytics/tagging/TagOrderCompletion.java
/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (C) Copyright IBM Corp. 2016
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 ******************************************************************************/
package com.tl.digitalanalytics.tagging;

import android.app.Activity;
import android.util.Log;

import com.da.kitchensink.R;
import com.digitalanalytics.api.DigitalAnalytics;
import com.tl.digitalanalytics.app.MainActivity;
import com.tl.digitalanalytics.model.Order;
import com.tl.digitalanalytics.model.Product;
import com.tl.digitalanalytics.model.ShoppingCart;

/**
 * 
 * @author Sohil Shah (sohishah@us.ibm.com)
 * 
 */
public class TagOrderCompletion {

	private ShoppingCart cart;
	private Order order;
	private Activity activity;

	public TagOrderCompletion(final Activity activity, ShoppingCart cart,
			Order order) {
		this.activity = activity;
		this.cart = cart;
		this.order = order;
	}

	public void executeTag() {
		try {

			Boolean success = true;

			// Setup Data to be tagged
			String orderId = this.order.getId(); // required
			String subtotal = this.order.getSubTotal(); // optional, null-able
			String customerId = this.order.getCustomerId(); // required
			String currency = this.order.getCurrencyCode(); // required

			// Fire a Shop Action9 for Product in the cart
			Product[] products = this.cart.getProducts();
			for (Product product : products) {
				String productId = product.getId(); // required
				String productName = product.getName(); // required
				String categoryId = product.getCategoryId(); // required
				String quantity = "" + product.getQuantity(); // optional,
																// null-able
				String baseUnitPrice = product.getBaseUnitPrice(); // optional,
																	// null-able
				String[] attributes = product.getAttributes(); // optional,
																// null-able

				success = DigitalAnalytics.fireShopAction9(productId, productName,
						quantity, baseUnitPrice, categoryId, orderId, subtotal,
						customerId, currency, attributes);
			}
			
			success = this.fireOrderTag();

			this.finish(success);

		} catch (Exception e) {
			Log.e(Constants.LOG_TAG, e.getMessage(), e);
			this.finish(false);
		}
	}

	private Boolean fireOrderTag() {
		String orderId = this.order.getId(); // required
		String subtotal = this.order.getSubTotal(); // optional, null-able
		String shippingCharge = this.order.getShippingCharge(); // optional,
																// null-able
		String customerId = this.order.getCustomerId(); // required
		String city = this.order.getCustomerCity(); // required
		String state = this.order.getCustomerState(); // required
		String zip = this.order.getCustomerZip(); // required
		String currency = this.order.getCurrencyCode(); // required
		String[] attributes = this.order.getAttributes();// optional, null-able

		// fire the order tag
		return DigitalAnalytics.fireOrder(orderId, subtotal, shippingCharge, customerId,
				city, state, zip, currency, attributes);
	}

    private void finish(boolean success) {
        // This is for functional testing/user feedback only. In practice,
        // tagging should just be silent
        Helper.myAlertDialog(activity, success, MainActivity.class, R.string.tag_order_placement_success, R.string.tag_order_placement_failed);

        // Clean up
        cart.clear();
    }
}
